File: MoneySplitter/Python/logic/calculator.py
"""Balance-calculation logic for the Money Splitter application."""

from typing import Dict, List
from data.models import TripData


def convert_amount(
    amount: float,
    from_currency: str,
    to_currency: str,
    base_currency: str,
    conversion_rates: Dict[str, float],
) -> float:
    """Convert *amount* between currencies via the trip's base currency as pivot.

    *conversion_rates* maps a non-base currency code to its base-currency
    equivalent (e.g. ``{"USD": 380, "EUR": 410}`` when base is HUF).
    The base currency itself is always implicitly 1.0.
    """
    if from_currency == to_currency:
        return amount

    # Step 1 – to base currency
    if from_currency == base_currency:
        amount_base = amount
    else:
        rate = conversion_rates.get(from_currency, 1.0)
        amount_base = amount * rate

    # Step 2 – from base currency to target
    if to_currency == base_currency:
        return amount_base

    rate = conversion_rates.get(to_currency, 1.0)
    return amount_base / rate if rate != 0 else 0.0


def calculate_balances(trip: TripData) -> Dict[str, float]:
    """Return the balance of every person in *trip.result_currency*.

    Positive  →  person is owed money (overpaid).
    Negative  →  person owes money (underpaid).
    """
    balances: Dict[str, float] = {name: 0.0 for name in trip.people}

    for row in trip.expenses:
        for col_idx, cell in enumerate(row):
            if cell.amount is None or cell.amount <= 0:
                continue
            if col_idx >= len(trip.people):
                continue

            payer = trip.people[col_idx]

            converted = convert_amount(
                cell.amount,
                cell.currency,
                trip.result_currency,
                trip.base_currency,
                trip.conversion_rates,
            )

            # Determine beneficiaries ----------------------------------
            beneficiaries: List[str] = [
                p for p in cell.checked_people if p in trip.people
            ]
            if not beneficiaries or set(beneficiaries) == set(trip.people):
                beneficiaries = list(trip.people)

            share = converted / len(beneficiaries)

            # Payer gets full credit; each beneficiary is debited.
            balances[payer] += converted
            for person in beneficiaries:
                balances[person] -= share

    return balances

File: MoneySplitter_APP/app/src/main/java/com/example/moneysplitter/logic/Calculator.kt
package com.example.moneysplitter.logic

import com.example.moneysplitter.data.Settlement
import com.example.moneysplitter.data.TripData
import kotlin.math.abs

object Calculator {

    fun convertAmount(
        amount: Double,
        fromCurrency: String,
        toCurrency: String,
        baseCurrency: String,
        rates: Map<String, Double>
    ): Double {
        if (fromCurrency == toCurrency) return amount

        // Step 1: convert to base currency (rates store: 1 FOREIGN = X BASE)
        val inBase = if (fromCurrency == baseCurrency) {
            amount
        } else {
            amount * (rates[fromCurrency] ?: 1.0)
        }

        // Step 2: convert from base to target
        return if (toCurrency == baseCurrency) {
            inBase
        } else {
            inBase / (rates[toCurrency] ?: 1.0)
        }
    }

    fun calculateBalances(trip: TripData): Map<String, Double> {
        val balances = mutableMapOf<String, Double>()
        trip.people.forEach { balances[it] = 0.0 }

        for (expense in trip.expenses) {
            if (expense.amount <= 0 || expense.paidBy.isBlank()) continue

            val convertedAmount = convertAmount(
                expense.amount,
                expense.currency,
                trip.resultCurrency,
                trip.baseCurrency,
                trip.conversionRates
            )

            val beneficiaries = expense.splitAmong.ifEmpty { trip.people }
            if (beneficiaries.isEmpty()) continue
            val share = convertedAmount / beneficiaries.size

            // Credit the payer
            balances[expense.paidBy] = (balances[expense.paidBy] ?: 0.0) + convertedAmount

            // Debit each beneficiary
            for (person in beneficiaries) {
                balances[person] = (balances[person] ?: 0.0) - share
            }
        }

        return balances
    }

    fun calculateSettlements(balances: Map<String, Double>): List<Settlement> {
        data class Entry(val name: String, var amount: Double)

        val debtors = mutableListOf<Entry>()
        val creditors = mutableListOf<Entry>()

        for ((person, balance) in balances) {
            when {
                balance < -0.01 -> debtors.add(Entry(person, abs(balance)))
                balance > 0.01 -> creditors.add(Entry(person, balance))
            }
        }

        debtors.sortByDescending { it.amount }
        creditors.sortByDescending { it.amount }

        val settlements = mutableListOf<Settlement>()
        var i = 0
        var j = 0

        while (i < debtors.size && j < creditors.size) {
            val amount = minOf(debtors[i].amount, creditors[j].amount)
            if (amount > 0.01) {
                settlements.add(Settlement(debtors[i].name, creditors[j].name, amount))
            }
            debtors[i].amount -= amount
            creditors[j].amount -= amount
            if (debtors[i].amount < 0.01) i++
            if (creditors[j].amount < 0.01) j++
        }

        return settlements
    }
}
